### DISCUSSION TITLE
Question regarding freezing layers

### DISCUSSION THREAD

**Discussion by ep0p** [mindee/doctr#1733] [Q&A] [answered]
I have  a question in regards to the layers freezing, since i stumbled on this suggestion a couple of times in the Q&As :

```
model = recognition.__dict__[args.arch](pretrained=args.pretrained, vocab=vocab)
# Freeze all layers except the last one
for name, param in model.named_parameters():
    if name in ["linear.weight", "linear.bias"]:  # specific layer for crnn_vgg16_bn
        param.requires_grad = True
        print(f"Set layer {name} to trainable")
        continue
    param.requires_grad = False

Currently we reset only the last linear layer (classifier) but let the whole model trainable if it's trained on a other vocab as french.
What you can do is modifying the training script slightly by freezing all layers up to the last layer.
In this case you would only train the classifier with your new vocab.
NOTE: This will only work for all crnn models and vitstr models because the other models (parseq, sar, master) contain embedding layer
```
Is this the same as the backbone-freezing or is it close, but not the same?

**Comment by felixdittrich92** [comment, marked as answer]
Hi @ep0p :wave:,

In the case of `backbone-freezing` we do really only freeze the feature extractor model part all other parts including the head are trainable ... in the snippet we freeze everything up to the head ..so only the last linear layer is trainable .. the layer names will change depending on the model / architecture :)

Personally i have never tried to tune only the head :sweat_smile:
